InformationScripting/src/queries/LinearQuery.h

```diff
@@ -0,0 +1,47 @@
+/***********************************************************************************************************************
+**
+** Copyright (c) 2011, 2015 ETH Zurich
+** All rights reserved.
+**
+** Redistribution and use in source and binary forms, with or without modification, are permitted provided that the
+** following conditions are met:
+**
+**    * Redistributions of source code must retain the above copyright notice, this list of conditions and the following
+**      disclaimer.
+**    * Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the
+**      following disclaimer in the documentation and/or other materials provided with the distribution.
+**    * Neither the name of the ETH Zurich nor the names of its contributors may be used to endorse or promote products
+**      derived from this software without specific prior written permission.
+**
+**
+** THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+** INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+** DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+** SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+** SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+** WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+** OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+**
+***********************************************************************************************************************/
+
+#pragma once
+
+#include "../informationscripting_api.h"
+
+#include "Query.h"
+
+namespace InformationScripting {
+
+class INFORMATIONSCRIPTING_API LinearQuery : public Query
+{
+	public:
+		virtual QList<TupleSet> execute(QList<TupleSet> input) override final;
+		virtual TupleSet execute(TupleSet) = 0;
+
+	protected:
+		template <class QueryClass>
+		using LinearExecuteFunction = std::function<TupleSet (QueryClass*, TupleSet)>;
```

Comment:
Hm that doesn't really work that simple. We would need to template the whole class, where the template arguments is the specific implementing class. This also is a bit complicated if a class is not a direct child class (e.g. AstQuery <: ScopedArgumentQuery <: LinearQuery).

The problem is that the exec_ function is just a std::function wrapper to a private member function and at creation you can't bind the instance (see AstQuery::registerDefaultQueries)

Any suggestion on how to solve this, since in general I like the idea.